procedures/automation-android-setup.md

```diff
@@ -0,0 +1,200 @@
+# How to set up automation for nightly Android product
+
+These instructions define which resources are needed for setting up nightly Android product release automation, as well as the steps to prepare the automation itself. Additionally, this guide include steps for running a "staging" version of the automation so that it can be tested in the developer's own repository.
+
+Throughout this document, wherever the term `$product` is used, substitute your product's name in (replacing spaces with hyphens), e.g.: `reference-browser` or `fenix`.
+
+Note: we don't need to explicitly define scopes in Taskcluster. We just define them in the task, hook's role, and each scriptworker script. Taskcluster will implicitly detect scopes, but it doesn't use them for anything other than generic equality/wildcard verification.
+
+1. Request dep and release signing keys (example bug for [`reference-browser`](https://bugzilla.mozilla.org/show_bug.cgi?id=1508761)). Confirm with the app's team what the "signature algorith" and "digest algorithm" should be, and include that information in the ticket.
+2. Clone the product's repository
+    * Note that, while doing "staging" runs, you'll have to use your `master` branch in your fork. Hook cloning/CoT/etc haven't been tested with a non-default branch
+3. Add `.taskcluster.yml` in the root of the repository if it doesn't exist, and add a task for cron jobs. ([`fenix` example, go to `tasks_for == "cron"`](https://github.com/mozilla-mobile/fenix/blob/master/.taskcluster.yml))
+    1. Update the `event.repository.clone_url` checks to use the correct repo URL
+    2. Update scopes to be relevant for your product
+    3. Update the apks in `in.payload.command` so that the paths match your product
+    4. Update `in.payload.image` so that a compatible docker image is used
+    5. Update `in.metadata` for your product
+4. In `automation/taskcluster`, add `decision_task_nightly.py`, `schedule_nightly_graph.py` and `lib/tasks.py` that will perform most of the automation wiring. ([`fenix` examples](https://github.com/mozilla-mobile/fenix/blob/master/automation/taskcluster/))
+    1. `schedule_nightly_graph.py` is mostly good as-is, but I'd recommend a once-over to make sure there's nothing Fenix-specific
```

Comment:
@tomprince, out of curiosity, does https://hg.mozilla.org/ci/taskgraph include a way of scheduling cron tasks? If not, I don't think this is a high-priority. I'm just curious.

Reply:
Arguably, that's not `taskgraph`'s responsibility, right? We have TC hooks which run the tasks according to cron. `schedule_nightly_graph` just invokes `.taskcluster.yml` with a bunch of Github environment variables set up

Reply:
taskgraph will have the same support for scheduling cron graphs (in conjunction with ci-admin):
- ci-admin creates a hook that has job that runs `taskgraph cron`
- `taskgraph cron` looks at `.cron.yml` in the appropriate repository and creates any jobs that are schedule then, using `.taskcluster.yml` for the task description.

It sounds like `schedule_nightly_graph` is analogous to `taskgraph cron` here.

Reply:
Sounds like I'm a little out-of-date here: I'm familiar with cron timings being set in the hook (which I suppose will be ci-admin'd). Once the job actually starts, it's just a regular hook trigger.

What kind of configuration is stored in `.cron.yml`? Does it change what jobs will run based on when a hook is triggered?

Reply:
@mitchhentges https://hg.mozilla.org/mozilla-central/file/tip/.cron.yml might be useful to glance at in this context. I'd assume it's going to be something similar.

Reply:
Wait, so we'll have _some_ cron-related build configuration in-repository, and some of it on our hooks (so, in ci-admin)? Doesn't this fragment our config a bit?

Reply:
For mozilla-central, the hooks are run every 15m, and then the task that runs is responsible for figuring out if there is anything scheduled at that time.

Reply:
Ohhh, interesting - would it be easier to have a separate hook for each build, so we don't need cron configuration in our repositories?

Reply:
I would be inclined to say it is easier the other way around: there is a standard hook that doesn't have any project specific details, and then everything can be configured from the repository, rather than needing to change hooks to get changed behavior.

Reply:
I feel like this solution is causing us to stack two solutions on top of each other where a single standalone one would work better.

For example, some downsides we're going to see are:
* artificial delays before builds start. We need to wait for the next 15 minute trigger from the _outside_ cron system before our _internal_ cron system can trigger builds (that potentially could've happened 14 minutes ago). Tweaking an arbitrary "poll" value isn't ideal
* We lose the benefits of the outside hook system:
    * I can't just "trigger" a hook if I need to run it early
    * I can't click "view last run" of a certain hook to see what went wrong
* We need to maintain our internal cron system
* Explaining the system to new relengers/devs: "You customize your cron job here in `.cron.yml`. No, our internal cron system doesn't actually start the build. Yeah, an entirely separate system that _also_ runs on a timer triggers our internal system, which delegates accordingly"
* FWIW, though minimal: when our 15m hook triggers, but no jobs need to run, it just uses unnecessary machine hours and populates our job-tracking systems with empty builds

Although, I can definitely see the benefit of having a `.cron.yml` in your repository: it's much better for reviewing changes and tracking history. Would our time making an internal cron system be better spent improving Taskcluster integration (e.g.: list each cron job separately as its own "virtual" hook?)